synapse/lib/storm.py

```diff
@@ -390,11 +396,50 @@ def loadRuntVars(self, query):
                 self.tick()
                 yield node, path
 
-    async def getScopeRuntime(self, query, opts=None):
+    async def getScopeRuntime(self, query, opts=None, impd=False, modl=False):
         runt = Runtime(self.snap, user=self.user, opts=opts)
+        runt.isModuleRunt = modl
+        runt.isImport = impd
+        if not impd:  # respect the import boundary
+
+            # if we are a top level module we need to
+            # push down our runtvars
+            if self.isModuleRunt:
+                for name in self.runtvars:
+                    if name not in self.vars:
+                        continue
+                    if name not in self.modulefuncs and name != 'lib':
+                        runt.globals.add(name)
+                        runt.runtvars.add(name)
+                        runt.vars[name] = self.vars[name]
+
+            # propagate down all the global variables
+            for name in self.globals:
+                if name not in self.modulefuncs and name != 'lib':
+                    runt.vars[name] = self.vars[name]
+                    runt.globals.add(name)
+                    runt.runtvars.add(name)
+
+            # regardless of module level, we have to reload the module functions
+            # we were given so they run with the right runt. We need this so we have
+            # the whole function telescoping going on, and module level variables get
+            # set to the right values
+            runt.modulefuncs = dict(self.modulefuncs)
+            for name, oper in self.modulefuncs.items():
+                runt.runtvars.add(name)
+                async for item in oper.run(runt, s_ast.agen()):
+                    pass  # pragma: no cover
+
         runt.loadRuntVars(query)
         return runt
 
+    async def propBackGlobals(self, runt):
+        for name in runt.globals:
+            if name in self.modulefuncs or name == 'lib':
+                # don't override our parent's version of a function
+                continue
+            self.vars[name] = runt.vars[name]
```

Comment:
This implies that the keys of runt.vars is a subset of runt.globals.  That **Should** be the case but a `.get(name, s_common.novalu)`, check and set can avoid a keyerror.

Reply:
:+1: